sdio: initial SDIO HAL implementation

rmsyn opened the pull request:
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [ ] I have updated existing examples or added new ones (if applicable).
- [ ] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [ ] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [ ] I have added necessary changes to user code to the [Migration Guide](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/MIGRATING-0.21.md).
- [ ] My changes are in accordance to the [esp-rs developer guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/DEVELOPER-GUIDELINES.md)

#### Extra:
- [ ] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

### Pull Request Details 📖

#### Description

Provides the initial skeleton for the `embedded_hal::mmc::MmcOps` implementation for the SDIO peripherals on the ESP32C6 microcontroller.


#### Testing

Doc tests for now.

Integration tests via `examples` program is WIP.

MabezDev commented on esp-hal/src/sdio/esp32c6.rs:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    /// Releases the inner SDIO peripherals back to the caller.
+    pub const fn release(self) -> (SLC, SLCHOST) {
```
FYI this pattern goes against our guidelines in esp-hal: https://github.com/esp-rs/esp-hal/blob/main/documentation/DEVELOPER-GUIDELINES.md.

We use peripheral singletons with lifetime parameters, you can "borrow" (our method is called `reborrow`, but now I'm wondering if `borrow` makes more sense cc: @bugadani) a peripheral to allow drivers to automatically "release" the peripheral singleton again once the driver has been dropped.

rmsyn replied:
I almost completely refactored using your advice, and the developer guidelines. Also took a look at some other peripheral implementations, and hopefully got closer to the intended style of `esp-hal`.

MabezDev commented on esp-hal/src/sdio/result.rs:
```diff
@@ -0,0 +1,39 @@
+//! Error and result types for SDIO peripherals.
+
+/// Convenience alias for the SDIO peripheral [Result](core::result::Result) type.
+pub type Result<T> = core::result::Result<T, Error>;
```
I don't think this is something that we want, if you need it for the eh impl then please inline it close to where the eh impl will live

rmsyn replied:
I removed the `Result` alias, but did keep the `Error` type. `Error` currently only holds one variant, but the plan is to implement more for specific failure conditions for the `hal::mmc` implementation.

MabezDev commented on esp-hal/src/sdio/esp32c6.rs:
```diff
@@ -0,0 +1,106 @@
+use esp32c6::{SLC, SLCHOST};
```
Don't use the pac directly, use it from `crate::peripherals` instead, this will allow you to target different chips in esp-hal later

rmsyn replied:
Thanks, I wasn't aware of the abstractions available, and replaced the direct use with `crate::peripherals`.

MabezDev commented on esp-hal/src/sdio.rs:
```diff
@@ -0,0 +1,12 @@
+//!
+//! The peripheral can be used to transfer data over the SDIO bus in `Slave` mode.
+
+#[cfg(feature = "esp32c6")]
```
It's likely this isn't needed based on my other comments.

rmsyn replied:
You were right, I refactored taking your comments into account, and this is no longer needed.

Dominaezzz commented on esp-hal/src/sdio.rs:
```diff
@@ -0,0 +1,208 @@
+
+impl core::error::Error for Error {}
+
+impl<'d> MmcOps for Sdio<'d> {
```
Something is not quite right here. `MmcOps` is meant for the master side of this protocol, but you've implemented this on the slave peripheral.

I think you want the SD/MMC peripheral (#1928) and not the SDIO peripheral, right?

rmsyn replied:
> Something is not quite right here. `MmcOps` is meant for the master side of this protocol, but you've implemented this on the slave peripheral.
> 
> I think you want the SD/MMC peripheral (#1928) and not the SDIO peripheral, right?

Correct, the `MmcOps` trait is currently for the host, and I need to split the trait into potentially three traits:

- `MmcCommon`
- `MmcHost`
- `MmcDevice`

I'll keep working on the e-h trait along with the work in this pull request. Thanks for the review.

rmsyn committed:
sdio: initial SDIO HAL implementation

Provides the initial skeleton for the `embedded_hal::mmc::MmcOps`
implementation for the SDIO peripherals on the ESP32C6 microcontroller.

Adds `Pins` + `Mode` type for SDIO GPIO pin configuration.

Adds the `State` type to represent the SDIO states + transitions.

Uses the experimental `embedded-hal-sdmmc` crate for SD/MMC traits.

See discussion for reasoning:

- <https://github.com/esp-rs/esp-hal/pull/3503#issuecomment-2921481055>
- <https://github.com/esp-rs/esp-hal/pull/3503#issuecomment-2922137021>
- <https://github.com/esp-rs/esp-hal/pull/3503#issuecomment-2923758921>

rmsyn commented:
Running the `CI / *` checks are failing, partially due to some weird dependency resolution.

As I get closer to a full implementation, I may need to bring in the `embedded-hal::mmc::*` stuff locally to avoid the git dependency.

What's strange is that running `cargo clippy ...`  directly passes, but the `xtask lint-packages` can't resolve the `embedded_hal::mmc::MmcCommon` trait, and fails to build the `embassy-hal-async` examples.

I'll keep debugging.

bugadani commented:
It would be best to not block this PR on embedded-hal, as it'll likely take months to stabilize a new set of traits. You should instead demonstrate new functionality in a new crate, like `embedded-sdmmc` in the embedded-hal repository, so that it can be released unstably and implemented by esp-hal and one other required implementor.

That way you wouldn't have to patch around the dependency issues with multiple embedded-hal versions in the same project.

MabezDev commented:
To expand on @bugadani's comment, we can't merge this with git dependencies, so I would take the suggestion to split this into a separate crate and _then_ work on upstreaming to eh proper. Feel free to iterate here in the mean time, but if you want this merged it won't be with a fork of embedded-hal.

rmsyn commented:
> It would be best to not block this PR on embedded-hal, as it'll likely take months to stabilize a new set of traits. You should instead demonstrate new functionality in a new crate, like embedded-sdmmc in the embedded-hal repository, so that it can be released unstably and implemented by esp-hal and one other required implementor.

This sounds like a good plan. I'll bring it up in next week's meeting, and see what the rest of the HAL team thinks.

> To expand on @bugadani's comment, we can't merge this with git dependencies, so I would take the suggestion to split this into a separate crate and then work on upstreaming to eh proper. Feel free to iterate here in the mean time, but if you want this merged it won't be with a fork of embedded-hal.

Right, I had no intention of trying to get this PR merged with the git dependency. My original thought process was to prove the implementation here and in `jh71xx-hal`, iterate on the trait in `embedded-hal`, stabilize in `embedded-hal`, and take this PR out of `Draft` status after an `embedded-hal` release that includes the `mmc` module + traits.

However, the solution discussed above sounds like a better plan. We could do a `embedded-hal-sdmmc` `v0.1.0-alpha.whatever` release, and base the work off that crate until `mmc` stuff gets merged upstream (if ever).

Without trying to bikeshed too much, we would probably need to call it `embedded-hal-sdmmc`, since @thejpster already has [embedded-sdmmc-rs](https://github.com/rust-embedded-community/embedded-sdmmc-rs). `embedded-hal-sdmmc` also keeps with the naming convention of `embedded-hal-[async, io, nb, ...]`.

rmsyn committed:
sdio: add `Sdio::bus_width` method

Updates the `embedded-hal-sdmmc` dependency, and adds the
`Sdio::bus_width` method to get the bus width of the SDIO peripheral.

rmsyn committed:
sdio: add interrupt types

Adds the `HostInterrupt` enum to represent device-to-host interrupt
variants.

rmsyn committed:
sdio: add init pin configuration

Adds the pin configuration for the `Sdio::init` HAL method.

Leaves the `sdio::Configuration` implementation for a future commit.

Leaves the `Sdio::hal_hw_init` implementation for a future commit.

rmsyn committed:
sdio: add low-level init

Adds the low-level init function to configure low-level SDIO registers.

rmsyn committed:
sdio: add HINF register block

Adds the types needed to represent the `HINF` register block for SDIO
peripherals.

rmsyn committed:
sdio: add config + enable highspeed

Adds a basic `Config` struct to represent SDIO peripheral configuration.

Adds a method to enable highspeed on the SDIO peripheral based on
configuration.

rmsyn committed:
sdio: add set timing method

Adds the `Timing` struct to represent SDIO timing settings.

Adds a method to set low-level timing configuration for the SDIO
peripheral.

rmsyn committed:
sdio: change interrupt types to bitfields

Changes the SDIO interrupt types to be bitfield structs.

Adds the `DeviceInterrupt` type to represents interrupts sent from the
host to the device.

rmsyn committed:
sdio: add low-level interrupt enable method

Adds a method to perform low-level interrupt enabling for the SDIO
device peripheral.

rmsyn commented:
I added the remaining functions needed for the SDIO `init` implementation.

Not sure about the best way to approach SOC-specific methods (i.e. the `low_level_*` methods). The methods do more-or-less the same thing on each SOC, but the register names are different, especially between the ESP32 and ESP32-C6 interrupt naming conventions.

One design I though of was to add a `SdioLowLevel` trait, and implement the trait on SOC-specific data structures, instead of the `Any<Peripheral>` types.

@Dominaezzz @MabezDev @bugadani thoughts?

Dominaezzz commented:
Regarding the low level stuff, see https://github.com/esp-rs/esp-hal/issues/1282 . Whilst esp-idf serves as a (mostly reliable) working reference, its design decisions (which were made for a C library) shouldn't be copied directly (into a Rust hal).

Check out what the stable(izing) drivers (SPI, I2C, UART, GPIO) are doing to get an idea of what to do. Also the [dev guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/DEVELOPER-GUIDELINES.md) is another place to check.

imo, for chip-based differences, `cfg`s are more appropriate than traits. traits are more appropriate  for peripheral-based differences on the same chip.

rmsyn commented:
> Whilst esp-idf serves as a (mostly reliable) working reference, its design decisions (which were made for a C library) shouldn't be copied directly (into a Rust hal).

Agreed, I kept the relative structure of the code the same. It should hopefully make reviewing the code easier, since the mapping from the Rust back to the `esp-idf` function names is pretty close. I obviously changed the functions to be idiomatic Rust.

> imo, for chip-based differences, cfgs are more appropriate than traits. traits are more appropriate for peripheral-based differences on the same chip.

This is the approach I ended up going with, which can hopefully be extended pretty easily as more SDIO-having SOCs are added.

bugadani commented on esp-hal/src/sdio.rs:
```diff
@@ -145,16 +151,31 @@ impl<'d> Sdio<'d> {
         self.hinf.info()
     }
 
+    /// Convenience function to get a reference to the SLCHOST register block.
```
copy-paste?

bugadani commented on esp-hal/src/sdio.rs:
```diff
@@ -192,16 +213,42 @@ impl<'d> Sdio<'d> {
+    }
+
+    /// Performs low-level initialization of the SDIO peripheral.
+    #[cfg(feature = "esp32")]
```
`cfg(esp32)`

bugadani commented on esp-hal/src/lib.rs:
```diff
@@ -235,6 +235,8 @@ pub mod i2c;
 pub mod peripheral;
 #[cfg(all(feature = "unstable", any(soc_has_hmac, soc_has_sha)))]
 mod reg_access;
+#[cfg(all(feature = "unstable", any(feature = "esp32", feature = "esp32c6")))]
```
```suggestion
#[cfg(all(feature = "unstable", any(sd_host))]
```

you'll need to fix up the metadata, if this works for the c6 we're missing the sd_host driver entry

rmsyn replied:
Can I use the `soc_has_sd` instead? Part of the `sdio` module will also include host drivers in the future, e.g. for the ESP32-P4.

Also, where are these `cfg` selectors defined? I see `soc_has_hmac` used, but not where it is defined.

bugadani replied:
it's defined in esp-metadata; each listed peripehral creates a soc_has_* symbol

bugadani commented:
If the difference is only register names, we should consider papering over that by patching the PAC.

bugadani commented on esp-hal/src/sdio.rs:
```diff
@@ -0,0 +1,586 @@
+//! # Secure Digital I/O - Slave Mode
```
This should be sdio/slave.rs

rmsyn replied:
Right, I think I copied this doc string from somewhere else. I'll move the current top-level stuff into a separate module.

Dominaezzz replied:
SDIO is always a slave driver.

- `src/sdio.rs`
- `src/sdmmc.rs`

bugadani replied:
Oki disregard me, I'm rather confused about the different names in this area, or even how the whole thing is supposed to work. But if it's always slave mode, then there's no reason to call out that it's slave mode.

bugadani commented on esp-hal/src/sdio/pins.rs:
```diff
@@ -0,0 +1,181 @@
+    ) -> Self {
+        Self {
+            mode,
+            clk_sclk: Output::new(clk_sclk, Level::Low, OutputConfig::default()),
+            cmd_mosi: Flex::new(cmd_mosi),
+            dat0_miso: Flex::new(dat0_miso),
+            dat1_irq: Flex::new(dat1_irq),
+            dat2: Flex::new(dat2),
+            dat3_cs: Flex::new(dat3_cs),
```
Why do you need to create pin drivers?

rmsyn replied:
Because, at least in `SDIO` mode, some of these pins act as both input and output. Is there a better way to express that using the `PeripheralInput/Output` types?

bugadani replied:
The API is meant to take `impl PeripheralInput` and `impl PeripheralOutput`. Internally, drivers are supposed to use gpio::interconnect types, namely `OutputSignal` and `InputSignal`. You can take a look at, say, I2C for an example.

bugadani replied:
Just noting that OutputSignal is a slight misnomer as it is an input/output signal in truth.

rmsyn replied:
After reviewing the available `InputSIgnal/OutputSignal` definitions in `esp-metadata`, there don't appear to be signal definitions for `SDIO_CLK`, `SDIO_CMD`, or `SDIO_DATA[0-3]` in the `esp32` or `esp32c6` TOML files. The ESP32-C6 TRM also doesn't have those signals defined. How should I proceed?

bugadani replied:
let us know if you can find this in esp-idf, otherwise we'll ask around internally

Dominaezzz replied:
There aren't signals for the SDIO, it's IO_MUX only, like the USB_OTG signals. Have a look at the USB driver for inspiration.

rmsyn replied:
I ended up finding the signals in the ESP32 config in `esp-metadata`. The config just leaves out the signal number, so that is what I did for the ESP32-C6 config. `SD_CMD` was also missing from the input signals for ESP32, so I added it since the command line is full duplex I/O.

bugadani replied:
OH

bugadani commented on esp-hal/src/sdio/pins.rs:
````diff
@@ -0,0 +1,181 @@
+    /// ```
+    pub fn new(
+        mode: Mode,
+        clk_sclk: impl OutputPin + 'd,
+        cmd_mosi: impl Pin + 'd,
+        dat0_miso: impl Pin + 'd,
+        dat1_irq: impl Pin + 'd,
+        dat2: impl Pin + 'd,
+        dat3_cs: impl Pin + 'd,
+    ) -> Self {
````
These should be PeripheralInput/PeripheralOutput

bugadani commented on esp-hal/src/sdio.rs:
```diff
@@ -0,0 +1,586 @@
+
+        match self.config.timing() {
+            Timing::PsendPsample => {
+                host.conf().modify(|_, w| unsafe {
+                    w.frc_sdio20().bits(0x1f);
+                    w.frc_sdio11().bits(0x0);
+                    w.frc_pos_samp().bits(0x1f);
+                    w.frc_neg_samp().bits(0x0)
+                });
```
I think only this part needs to be chip specific, not the outer logic.

rmsyn replied:
Right, so instead create eight new functions guarded by conditional compilation? I think your suggestion to fix register naming in the PAC is probably the best route to take.

bugadani replied:
What do you mean by 8? These 4 branches could be calls to the same function with 4 different sets of parameters (or a struct if you want them named), you need 2 variants that set the 4 register fields. That is one function per MCU, and 4 calls.

rmsyn replied:
> What do you mean by 8? These 4 branches could be calls to the same function with 4 different sets of parameters (or a struct if you want them named), you need 2 variants that set the 4 register fields. That is one function per MCU, and 4 calls.

Right, that's what the edited reply means. Just making two functions guarded by conditional compilation, and call the function four times. My initial response was mistaken.

bugadani commented on esp-hal/src/sdio.rs:
```diff
@@ -0,0 +1,586 @@
+pub use timing::Timing;
+
+/// SDIO peripheral instance.
+pub trait PeripheralInstance: crate::private::Sealed {
```
I'd prefer not having this trait, it's just extra indirection in the code that doesn't really do anything for us, it seems.

bugadani replied:
I2C is probably the best one right now, but none of our drivers need to use multiple IPs at the same time. I think you might have looked at SPI which currently has an absolutely redundant PeripheralInstance trait that is being cleaned up in one of my draft PRs :)

rmsyn commented on esp-hal/src/sdio/hinf.rs:
```diff
@@ -0,0 +1,47 @@
+
+unsafe impl Sync for HinfInfo {}
+
+impl PeripheralInstance for AnyHinf<'_> {
```
Could you explain more, I'm not sure what you mean? I was working off the implementations I saw in other peripherals, so I am probably missing something. I copied the `impl PeripheralInstance` for `AnyHinf` to the inner `HINF` because the doctest complained about the `HinfInstance` trait not being implemented for the inner instance. From your comment, I'm guessing this was the wrong way to solve that problem?

bugadani replied:
Look at how instance traits are implemented for AnyI2c or AnySpi.

rmsyn committed:
metadata: add missing `SDIO_*` signals for ESP32-C6

Adds the missing `SDIO_*` GPIO signals for the ESP32-C6 device.

rmsyn committed:
sdio: refactor pins to use `OutputSignal`

Refactors the `sdio::Pins` type to use the `gpio::PinGuard` +
`gpio::interconnect::{InputSignal, OutputSignal}` types instead of the
lower-level `Pin` types.

bugadani commented on esp-metadata/devices/esp32c6.toml:
```diff
@@ -251,6 +252,12 @@ input_signals = [
     { name = "SPIHD",               id = 123 },
     { name = "SPIWP",               id = 124 },
+
+    { name = "SD_CMD" },
+    { name = "SD_DATA0" },
+    { name = "SD_DATA1" },
+    { name = "SD_DATA2" },
+    { name = "SD_DATA3" },
```
These should follow the naming in the TRM, i.e. `SDIO_*`. I'm adding these signals among others in a separate PR, but in case you win that race, please rename these :)

rmsyn replied:
I rebased on your changes, and updated the names in ESP32 to be consistent with ESP32C6. This just makes it easier to use in the SDIO driver, let me know if you want me to do something different.

rmsyn commented:
I'm starting to implement the `read_command/write_response`, and the `read/write_data` functions. From the TRM, it looks like the SDIO DMA engine is a separate, dedicated DMA distinct from the GDMA engine used for other peripherals.

Even the DMA descriptors are slightly different from the GDMA/PDMA descriptors. I'm thinking of adding a new `sdio_dma` (not sure of naming) module in the existing `dma` module. Since the DMA work is somewhat standalone, I'm thinking it may be good to put that work into a separate PR, as this one is already quite large. What do you all think?

bugadani commented:
> What do you all think?

Sounds good

rmsyn committed:
dma: add `Owner` bool conversion

rmsyn committed:
sdio: add DMA descriptor types

Adds DMA descriptor types based on the main DMA engine descriptors.

The main difference is the bitfield layout used in the descriptor flags.

rmsyn committed:
dma: better `Owner` ergonomics

Adds some methods to the `Owner` type for better ergonomics.

rmsyn committed:
dma: implement DescriptorFlagFields

Implements the `DescriptorFlagFields` trait for the relevant types.

Small refactors to handle function naming differences.

rmsyn committed:
dma: add DescriptorFlagFields trait

Adds the `DescriptorFlagFields` trait to generalize field accessor
functions for descriptor flag types.

Enables using a generic parameter for `DmaDescriptor`.

rmsyn commented:
@bugadani something broke with the new peripheral generation. `crate::peripherals::SLC` is no longer being generated on new builds. Any idea how I can fix this?

rmsyn commented:
@bugadani @Dominaezzz @MabezDev I made some changes to `DmaDescriptor` to make it generic over descriptor flag types.

There are still changes required to the traits + types that use `DmaDescriptor`, but I wanted to get your feedback on the approach. My thinking is that the memory layout is the same, with the exception of the bitfield in the descriptor flags, and making `DmaDescriptorGeneric` allows for more code re-use.

Let me know what you think.

rmsyn committed:
dma: make DmaDescriptor generic

Refactors the `DmaDescriptor` type to `DmaDescriptorGeneric` taking a
generic parameter for the flags field.

Creates convenience aliases for the concrete types used in the general +
SDIO dedicated DMA descriptors.

bugadani commented:
> Formatting fixes to make the `for_each_peripheral` macro more
readable/maintanable.

Please don't hand-edit generated code, it'll be nuked pretty quickly. If you know a way to format the generated code better than is done currently, do let us know.

rmsyn commented:
> Please don't hand-edit generated code, it'll be nuked pretty quickly. If you know a way to format the generated code better than is done currently, do let us know.

How are you currently formatting the generated code? I ran `cargo xtask format-packages` from the root, but nothing happened in `esp-metadata-generated`. Are you running `cargo +nightly fmt` in the generated output directory inside `target`.

Regardless, something is definitely broken if the code in `esp-metadata-generated` is meant to be updated mechanically. I just spent 3-4 hours figuring that shit out.

bugadani commented:
> How are you currently formatting the generated code?

`cargo xtask update-metadata` first runs the string through prettyplease, then calls rustfmt on it. Rustfmt is supposed to pick up settings from `rustfmt.toml` but is generally not very capable at formatting macros I think.

> Regardless, something is definitely broken if the code in esp-metadata-generated is meant to be updated mechanically. I just spent 3-4 hours figuring that shit out.

What problems are you trying to solve?

rmsyn commented:
> What problems are you trying to solve?

Automatically updating the code in `esp-metadata-generated` from the output of `esp-metadata`. That's what caused my builds to fail, sent me on a search for what changed, and took me hours to figure out.

If `cargo xtask update-metadata` is meant to update the code in `esp-metadata-generated` from the code in `esp-metadata`, it's not doing that. If that is not part of the xtask, it should be. For nothing else than to save other contributors the headache I just went through.

rmsyn commented:
> cargo xtask update-metadata first runs the string through prettyplease, then calls rustfmt on it. Rustfmt is supposed to pick up settings from rustfmt.toml but is generally not very capable at formatting macros I think.

The output is very hard to read, and even more difficult to update manually. It also breaks on random delimiters, but I'm not sure there is a good solution. I can look at `rustfmt.toml` and `prettyplease` docs to see if there are options to ignore line length, which would at least maybe help with breaking in random places.

It may be best to just add some lines to the `README` about how to update `esp-metadata-generated`. Something like "manually copy over code from `esp-metadata` output, run `cargo xtask update-metadata`, etc. I know it would have helped me quite a bit.

rmsyn commented:
> > If cargo xtask update-metadata is meant to update the code in esp-metadata-generated from the code in esp-metadata, it's not doing that
> 
> What do you mean it's not doing that? It works exactly as it should - you make changes to `esp-metadata`, run the command, get new code in `esp-metadata-generated`. If you don't get any changes in the generated code, you made changes that were ignored by esp-metadata. Here's a demo of me trying to add SLC to ESP32-C6:

Alright, I just ran `cargo xtask update-metadata` locally, and it works like you describe. I just missed the step to run it locally, because I didn't know it existed.

The CI was just failing at a different point, so tracking down the actual issue was not obvious.

rmsyn committed:
dma: make Preparation generic

Makes the `Preparation` type generic over the `DescriptorFlagFields`
trait.

rmsyn committed:
dma: make DmaLoopBuf generic

Maks `DmaLoopBuf` generic over the `DescriptorFlagFields` trait.

rmsyn committed:
dma: make DmaRxStreamingBuf generic

Makes the `DmaRxStreamingBuf` type generic over the
`DescriptorFieldFlags` trait.

rmsyn committed:
dma: make DescriptorSet generic

Makes the `DescriptorSet` type generic over the `DescriptorFlagFields`
trait.

rmsyn committed:
dma: make DmaTxBuf generic

Makes the `DmaTxBuf` generic over the `DescriptorFlagFields` trait.

rmsyn committed:
dma: make DmaRxBuf generic

Makes the `DmaRxBuf` generic over the `DescriptorFlagFields` trait.

rmsyn committed:
dma: make DmaRxTxBuf generic

Makes the `DmaRxTxBuf` type generic over the `DescriptorFlagFields`
trait.

rmsyn committed:
dma: make ChannelTx generic

Makes the `ChannelTx` struct generic over the `DescriptorFlagFields`
trait.

rmsyn commented:
> The generics are rather infectious but it is what it is.

I noticed that, too. Part of why I wanted to get feedback before fully changing over all of the needed types in the DMA module.

> I'm slightly anxious about making all these DMA structures generic for just the C6's and 32's SDIO peripheral.
> I also don't want to set a precedent for other peripheral specific DMAs like RMT, Ethernet, VDMA, 2D-DMA for other people to try and shoehorn those in with generics.

I'm not sure what the best alternative would be. There are already a number of traits in the `dma` module, and making a high-level DMA trait to suit all the use-cases would take a non-trivial amount of design work (probably outside the scope of this PR).

I'd be willing to work on the changes necessary for a truly generic DMA.

Part of the reason for making the type generic over the descriptor flags was that the bitfield layout was the only difference.

I would need to look into the other DMA types you listed to see if they share a similar layout. One solution could be to introduce a marker trait, or additional constraints to the `DescriptorFlagFields` trait.

Another solution could be to add some documentation to describe the memory layout of the DMA linked list, and say "if your peripheral's dedicated DMA engine shares this memory layout, implement these traits over its descriptor flags type, otherwise implement inside a `dma` submodule in your peripheral module".

I definitely don't want to make the mistake of trying to shove too many things into one box, but I would like to maximize code re-use if all of the DMA structures share a similar-enough memory layout.

The specifics of how the DMA engines access the linked lists of decriptors and buffers may also require unique implementations.

Dominaezzz commented:
> I'd be willing to work on the changes necessary for a truly generic DMA.

No don't worry about this. The other DMAs look wildly different. SDIO is the only one that looks similar.

Name wise I think I'd prefer, `struct DmaDescriptor<Flags = DmaDescriptorFlags>`, emphasis on the default generic param instead of the `Generic` suffix

rmsyn committed:
dma: add generic parameter to ChannelRx

Adds a generic paramter for the descriptor flags for the `ChannelRx`
type.

rmsyn committed:
dma: make DmaTxBuffer generic

Makes the `DmaTxBuffer` generic over the `DescriptorFlagFields` trait.

rmsyn committed:
dma: make DmaRxBuffer generic

Makes the `DmaRxBuffer` trait generic over the `DescriptorFlagFields`
trait.

rmsyn committed:
sdio: add DMA atomic descriptor

Adds the `AtomicDmaDescriptor` type to enable safely storing DMA
descriptors in static variables.

rmsyn committed:
sdio: add atomic buffer

Adds the `AtomicBuffer` type to allow storing byte buffers safely in
static variables.

rmsyn commented:
I added an initial integration test for `sdio` to test out some ideas for the static DMA structures.

It looks like `esp32` fails with:

```
Frame 0: _DoubleExceptionVector @ 0x400803c0> @ 0x00000000400803c0
```

and `esp32c6` fails with:

```
 panicked at /home/runner/work/esp-hal/esp-hal/esp-hal/src/gpio/interconnect.rs:475:9:
SDIO_CLK cannot be routed through the GPIO matrix
```

Not sure where the first error originates from, and don't really understand how to address the second.

I'll address the CI failures in another push of changes.

rmsyn commented:
> https://github.com/esp-rs/esp-hal/pull/3840

Thank you, that fixed the issue with `SDIO_CLK`. The test panics now because the `SDIO_*` signals are not tied to an `id` in the `esp-metadata`, and the indexing is causing an out-of-bounds error.

For `esp32c6`, it looks like there is a gap between `105` and `114` that could fit the `SDIO_*` signals. However, there doesn't appear to be an explicit mapping in the technical reference manual.

For `esp32`, there are `gpio_sd0-sd7` signals defines for indices `100-107`, but I think these just correspond to SDIO data channels? Not sure where to map the corresponding input channels, or the signals for `SDIO_[CLK, CMD]`. Similar to `esp32c6`, there are gaps in the GPIO signal mapping that could fit the required signals.

What are your thoughts on deviating from the TRMs, and creating a pseudo-mapping for the necessary signals? Would that even work, is more plumbing needed to handle unmapped signals?

rmsyn commented:
> What are your thoughts on deviating from the TRMs, and creating a pseudo-mapping for the necessary signals? Would that even work? Is more plumbing needed to handle unmapped signals?

I added a commit with re-definitions of the SDIO signals to previously unmapped signal numbers. This feels hacky, so please let me know if you would prefer I take a different approach.

bugadani commented:
> This fixed the esp32c6 error!

Awesome!

It's fine to ignore ESP32 for now, we can figure that chip out once C6 works as you want it to. Until then it's fine to remove the test for it, if it'll just double except anyway.

Tests otherwise tend to fail from time to time, for one reason or another. No matter.

rmsyn commented:
> It's fine to ignore ESP32 for now, we can figure that chip out once C6 works as you want it to. Until then it's fine to remove the test for it, if it'll just double except anyway.

Ok, I was digging through how to hook up a JTAG probe to the `esp32`, but I'll refocus on the `esp32c6` tests since the setup is much simpler.

> Tests otherwise tend to fail from time to time, for one reason or another. No matter.

For sure, I won't worry about it then.

rmsyn commented:
So, the HIL SDIO test for ESP32C6 works with the SPI pins hooked up with jumpers to the corresponding SDIO pins. In my local tests, I physically connected them. Will something similar be needed in the CI HIL test environment? The tests still need work to actually verify the transfers, and I'll be adding more tests for a full SDIO transaction following the protocol specified in the TRM.

Thanks again for all the help along the way!

rmsyn committed:
dep: add `bitfielder` dependency

Adds the `bitfielder` library as a more performant and ergonomic
bitfield library.

rmsyn committed:
WIP: sdio: add raw command/response functions

Adds `Sdio::read_command_raw` + `Sdio::write_response_raw` functions to
read and write raw data on the `CMD` line.

TODO: add SDIO mode implementations

NOTE: these functions may be unecessary, the controller seems to handle
command/response in hardware (at least in SPI mode).

rmsyn committed:
sdio: add command types

Adds SDIO commands supported by ESP SDIO controllers.

rmsyn committed:
sdio: qa-test integration tests

Adds a split `qa-test` SPI-mode SDIO integration test.

The tests are split with the intention to flash each half of the test to
separate devices, allowing for concurrent execution.

rmsyn committed:
sdio: add response types

Adds the SD and SPI mode `R5` response types, and the corresponding
field types.

rmsyn commented:
Apologies for stalling on this, I was waiting on 10kOhm resistors to arrive, and they did last week. I added them to the bench circuit, but they didn't help. The device still returns garbage to the host using SDIO in SPI-mode.

I'm honestly at a loss for what else to even try. The communication at this phase of the SDIO protocol is all happening in hardware in the SPI and SDIO DMA engines, so it feels like even hooking something like `probe-rs` into the loop won't do much good. If anyone has other ideas on how to continue debugging this issue, I'm willing to try whatever approach looks most promising.

rmsyn committed:
sdio: rebase on esp-hal 1.0.0

rmsyn committed:
fixup: qa-test: better handling of SDIO protocol

Initializes the SDIO device in the integration test.

Removes the `CMD5` sequence from the SDIO host integration test.

TODO: add interrupt handler for SDIO interrupts

rmsyn commented:
Rebased on the latest `main` (`d9815b205`) for the `1.0.0` release changes. Hopefully, shouldn't need any further rebases.

Cleaned up the commit history to remove `fixup` commits, and squashed the first three commits to get rid of the changes that included the `embedded-hal` fork.

The `qa-test` integration tests are reliably performing bit transfers again, though I think I need to implement SDIO interrupt handlers to properly follow the ESP SDIO protocol.

Should I use `SoftwareInterrupt`, or is there a configured hardware interrupt for SDIO? Asking specifically about configuring the `handler`, since the SDIO interrupt sources are all in registers.

It's also unclear to me if the SDIO device sends the interrupt over the wire, e.g. in SPI mode. The `qa-test` integration tests are running on separate devices, so they can't access the source interrupt registers directly (obviously). My assumption is that the source interrupt register is set, e.g. device interrupts host, some hardware magic happens, the interrupt is transferred over the wire, some hardware magic happens on the host, the host hardware interrupt handler fires. Does this sound correct?

hydra commented:
@rmsyn and chance you could update this PR?  I'm looking to use an SD card in my ESP-C6 project.
